lib/syscall_shim: Autogenerate `uk_syscall_*` symbols at build time

mogasergiu opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Introduce a new AWK script that autogenerates an additional, custom 
defined, exportsyms that contains all `uk_syscall_*` symbols as per 
each library's `UK_SYSCALLS_PROVIDED-*`.                            
                                                                    
This will help in reducing exportsyms.uk noise from now on as we no 
longer have to manually add these supposedly internal symbols.      

Thus, also remove such no longer necessary symbols.

This was done with the Introduction of `EACHOLIB_EXPORTS`,
the equivalent of `EACHOLIB_LOCALS`, but for exportsyms.uk.  
Now, one is able to add custom defined/named exportsyms.uk, in addition
to an already existing exportsyms.uk. If no exportsyms.uk exists, then,
unlike `EACHOLIB_LOCALS` which is processed regardless of whether a    
localsyms.uk exists or not, `EACHOLIB_EXPORTS` will be ignored.        
However, its result files may still be built as per Makefile recipe    
dependencies.

mogasergiu committed:
Introduce `EACHOLIB_EXPORTS`

Introduce the equivalent of `EACHOLIB_LOCALS`, but for exportsyms.uk.
Now, one is able to add custom defined/named exportsyms.uk, in addition
to an already existing exportsyms.uk. If no exportsyms.uk exists, then,
unlike `EACHOLIB_LOCALS` which is processed regardless of whether a
localsyms.uk exists or not, `EACHOLIB_EXPORTS` will be ignored.
However, its result files may still be built as per Makefile recipe
dependencies.

Checkpatch-Ignore: UNIKRAFT_SUBJECT_FORMAT
Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim: Autogenerate `uk_syscall_*` symbols at build time

Introduce a new AWK script that autogenerates an additional, custom
defined, exportsyms that contains all `uk_syscall_*` symbols as per
each library's `UK_SYSCALLS_PROVIDED-*`.

This will help in reducing exportsyms.uk noise from now on as we no
longer have to manually add these supposedly internal symbols.

Checkpatch-Ignore: SPDX_LICENSE_TAG
Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib: Remove unneeded `uk_syscall_*` symbols from all `exportsyms.uk`

The syscall shim library is now able to automatically generate and
export `uk_syscall_*` symbols by preprocessing each library's defined
`UK_SYSCALLS_PROVIDED-*`. Thus, we no longer need to manually add these
in each exportsyms.uk.

Note that in the case of `posix-futex`'s library we had to keep an
exportsyms containing `none` to still avoid having any symbols being
spilled.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/posix-process: Remove leftover `uk_syscall_` symbols of `clone3`

Commit cee6ce09d720 ("lib/posix-process: remove clone3 syscall")
tried to remove `clone3` but forgot to remove its `uk_syscall_` symbols
from `exportsyms.uk`.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

michpappas commented on lib/posix-futex/exportsyms.uk:
```diff
@@ -1,4 +1 @@
-uk_syscall_r_futex
-uk_syscall_e_set_tid_address
-uk_syscall_r_set_tid_address
+none
```
This exports a symbol called `none`, so what was done in `posix-futex` is definitely a hack to work around some bug in the build system. Do you think it's something easy to fix instead of propagating this hack? What I had in mind was wrapping calls to objcopy around a function that first checks if `exportsyms.uk` is present.

mogasergiu replied:
Opened an issue for this so that we can look at it later.

mogasergiu replied:
Pushed an update that uses `# No public symbols` instead of `none` to be more explicit and consistent.

mogasergiu replied:
So I tried allowing empty `exportsyms.uk` to exist
```diff
diff --git a/support/build/Makefile.rules b/support/build/Makefile.rules
--- a/support/build/Makefile.rules
+++ b/support/build/Makefile.rules
@@ -1024,8 +1024,13 @@ $(call libname2olib,$(1)): $(call libname2preolib,$(1)) \
                $$($(call vprefix_lib,$(1),LOCALS)) $$($(call vprefix_lib,$(1),LOCALS-y)) \
                $$(EACHOLIB_LOCALS) $$(EACHOLIB_LOCALS-y)
     $(call build_cmd,OBJCOPY,,$(call libname2olib,$(1)),\
-        $(OBJCOPY) $$(addprefix --keep-global-symbols=,$$($(call vprefix_lib,$(1),EXPORTS)) $$($(call vprefix_lib,$(1),EXPORTS-y))) \
-               $$(addprefix --localize-symbols=,$$($(call vprefix_lib,$(1),LOCALS)) $$($(call vprefix_lib,$(1),LOCALS-y))) \
+        $(OBJCOPY) \
+               $$(for f in $$($(call vprefix_lib,$(1),EXPORTS)) $$($(call vprefix_lib,$(1),EXPORTS-y)); do \
+                [ -s "$$$$f" ] && echo --keep-global-symbols="$$$$f"; \
+               done) \
+               $$(for f in $$($(call vprefix_lib,$(1),LOCALS)) $$($(call vprefix_lib,$(1),LOCALS-y)) $$(EACHOLIB_LOCALS) $$(EACHOLIB_LOCALS-y); do \
+                [ -s "$$$$f" ] && echo --localize-symbols="$$$$f"; \
+               done) \
                $$(addprefix --localize-symbols=,$$(EACHOLIB_LOCALS) $$(EACHOLIB_LOCALS-y)) \
                    $$(OBJCFLAGS) $$(OBJCFLAGS-y) \
                    $$($(call vprefix_lib,$(1),OBJCFLAGS)) $$($(call vprefix_lib,$(1),OBJCFLAGS-y)) \
```
while this works for not making objcopy error out on empty files given an argument to `--keep-global-symbols`/`--localize-symbols`, it is not enough to ensure that no symbols get exported. What we have to do is to essentially create the equivalent of a `localsyms.uk` for all public symbols of that library.

Therefore, since this is seemingly not just a quick fix and may require some additional thought, I have opened an internal issue for this in the backlog.

mogasergiu commented:
Rebase force push

mogasergiu commented:
Rebase push

mogasergiu committed:
lib/ramfs: Explicitly mention that there are no public symbols

`exportsyms.uk` of RAMFS had a `none` dummy string so that no symbols
are exported, otherwise the `objcopy --keep-global-symbols=` would have
failed due to an empty file being given as argument.
However, this is inconsistent with other such files which just had
`# No public symbols` which is just as effective. Thus make it
consistent and do the same here.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

michpappas commented:
Very cool to see so much noise symbols go away 🚀 

Reviewed-by: Michalis Pappas <michalis@unikraft.io>
Approved-by: Michalis Pappas <michalis@unikraft.io>
